### PR TITLE
[CBRD-24251] Reuse ODBC connection in CAS gateway

### PR DESCRIPTION
**Pull request by airnet73**
http://jira.cubrid.org/browse/CBRD-24251

**Purpose**
mysql server when the query is finished.
If you look at the session state with netstat when the connection to the mysql server is disconnected, you can see that the TIME_WAIT state is changed to the CLOSE state after a certain period of time has elapsed.
If there are more connections than the speed of switching from TIME_WAIT to CLOSE, it becomes impossible to connect to the mysql server due to insufficient ports.

**Implementation**
Do not repeat "connection -> close" whenever a query is executed in cgw, and if there is a reusable session, reuse it.
* When a connection request is made, it is checked whether there is an already connected session.
* If there is no connection, try to connect.
* If there is a session, it is determined whether reuse is possible. If reuse is not possible, a connection is attempted.

**Remarks**
N/A

### CODE REVIEW

File: src/broker/cas_cgw.c [hunk 1]
```diff
@@ -163,22 +172,14 @@ cgw_database_connect (SUPPORTED_DBMS_TYPE dbms_type, const char *connect_url)
 						out_connect_str,
 						(SQLSMALLINT) sizeof (out_connect_str),
 						&out_connect_str_len, SQL_DRIVER_NOPROMPT));
-
+      is_connet = true;
```
Comment (kisoo-han) [inline src/broker/cas_cgw.c]: What about use more meaningful name:
**is_connected**

Comment (airnet73) [reply src/broker/cas_cgw.c]: Thanks for your comments. Changed to is_connected.

**Review by kisoo-han — APPROVED** [review]

File: src/broker/cas_execute.c [hunk 2]
```diff
@@ -1064,7 +1062,7 @@ ux_cgw_prepare (char *sql_stmt, int flag, char auto_commit_mode, T_NET_BUF * net
       goto prepare_error;
     }
 
-  err_code = cgw_get_handle (&srv_handle->cgw_handle, true);
+  err_code = cgw_get_handle (&srv_handle->cgw_handle);
```
Comment (beyondykk9) [inline src/broker/cas_execute.c:1065]: no necessary to check cgw_handle and cgw_handle are null?

File: src/broker/cas_cgw.c [hunk 3]
```diff
@@ -1025,14 +1019,31 @@ cgw_error_msg (SQLHANDLE hHandle, SQLSMALLINT hType, RETCODE retcode)
 void
 cgw_database_disconnect ()
 {
-  cgw_cleanup_handle (local_odbc_handle);
+  if (local_odbc_handle == NULL)
+    {
+      return;
+    }
+
+  if (local_odbc_handle->hstmt)
```
Comment (byungwook-kim) [inline src/broker/cas_cgw.c:1027]: how about memset( local_odbc_handle, 0, sizeof(local_odbc_handle) );

File: src/broker/cas_cgw.c [hunk 4]
```diff
@@ -1025,14 +1019,31 @@ cgw_error_msg (SQLHANDLE hHandle, SQLSMALLINT hType, RETCODE retcode)
...
+    {
+      SQLDisconnect (local_odbc_handle->hdbc);
+      SQLFreeHandle (SQL_HANDLE_DBC, local_odbc_handle->hdbc);
+      local_odbc_handle->hdbc = NULL;
+    }
 
-  is_database_connected = -1;
+  connected_db_name[0] = '\0';
+  connected_db_user[0] = '\0';
+  connected_db_passwd[0] = '\0';
```
Comment (byungwook-kim) [inline src/broker/cas_cgw.c:1042]:
```suggestion
  connected_db_name[0] = '\0';
  connected_db_user[0] = '\0';
  connected_db_passwd[0] = '\0';
```
how about memset( local_odbc_handle, 0, sizeof(local_odbc_handle) );

Comment (byungwook-kim) [inline src/broker/cas_cgw.c:1042]:
```suggestion
  memset( local_odbc_handle, 0, sizeof(local_odbc_handle) );
```

File: src/broker/cas_cgw.c [hunk 5]
```diff
@@ -1123,6 +1128,9 @@ cgw_set_bindparam (T_CGW_HANDLE * handle, int bind_num, void *net_type, void *ne
 	size_t precision, scale;
 	char num_str[64];
 	char tmp[64];
+	SQLHDESC hdesc = NULL;
+
+	SQL_CHK_ERR (handle->hdbc, SQL_HANDLE_DBC, err_code = SQLAllocHandle (SQL_HANDLE_DESC, handle->hdbc, &hdesc));
```
Comment (beyondykk9) [inline src/broker/cas_cgw.c:1133]: no necessary to check hdesc is null?

File: src/broker/cas_execute.c (hunk 2 above)
Comment (airnet73) [reply src/broker/cas_execute.c:1065]: The cgw_get_handle function allocates a cgw handle to srv_handle->cgw_handle.
And if cgw_handle is null, it returns an error.

File: src/broker/cas_cgw.c (hunk 5 above)
Comment (airnet73) [reply src/broker/cas_cgw.c:1133]: If assignment from SQLAllocHandle to desc fails, an error occurs. At this time, it moves to goto to handle the error.

**Review by beyondykk9 — APPROVED** [review]

**Pull request merged**